From 9c2f5f23865ee1b83f79a08b21a54bfcad8a5af6 Mon Sep 17 00:00:00 2001
From: mlugg <mlugg@mlugg.co.uk>
Date: Wed, 24 Sep 2025 00:36:17 +0100
Subject: [PATCH 119/133] cbe: workaround GCC miscompilation
X-Developer-Signature: v=1; a=openpgp-sha256; l=2185; i=xtex@astrafall.org;
 h=from:subject; bh=qHy2ZbddKeVssc+dyf5Ffz6q86PylVHRZPuZ9smblI4=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78viW1ixU/tVmfkud1atj14pUDfc9kb+3uJh+szq9q
 eC9T156x0QWBjEuBksxRZYiwwZvVp10ftFl5bIwc1iZQIZIizQwMDAwsDDw5SbmlRrpGOmZahvq
 GRrqGOkYMXBxCsBUSxUwMlz4tuPElyeLj3703Mkpu8lO9mjW7OP8F1rE111tW3Tt8evtDP/9uD8
 x/62aYOo1q4zBw+K8eGKrYtKHLz9WcC1xqVJi3ssCAA==
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

This was causing a zig2 miscomp, which emitted slightly broken debug
information, which caused extremely slow stack unwinding. We're working
on fixing or reporting this upstream, but we can use this workaround for
now, because GCC guarantees arithmetic signed shift.

(cherry picked from commit 3f84b6c80ed3306f040dd98b8ccba561a052167a)
---
 lib/zig.h    | 8 ++++++++
 stage1/zig.h | 8 ++++++++
 2 files changed, 16 insertions(+)

diff --git a/lib/zig.h b/lib/zig.h
index 5c96b4bea0d3..5253912490ce 100644
--- a/lib/zig.h
+++ b/lib/zig.h
@@ -1510,8 +1510,16 @@ static inline zig_u128 zig_shl_u128(zig_u128 lhs, uint8_t rhs) {
 }
 
 static inline zig_i128 zig_shr_i128(zig_i128 lhs, uint8_t rhs) {
+    // This works around a GCC miscompilation, but it has the side benefit of
+    // emitting better code. It is behind the `#if` because it depends on
+    // arithmetic right shift, which is implementation-defined in C, but should
+    // be guaranteed on any GCC-compatible compiler.
+#if defined(zig_gnuc)
+    return lhs >> rhs;
+#else
     zig_i128 sign_mask = lhs < zig_make_i128(0, 0) ? -zig_make_i128(0, 1) : zig_make_i128(0, 0);
     return ((lhs ^ sign_mask) >> rhs) ^ sign_mask;
+#endif
 }
 
 static inline zig_i128 zig_shl_i128(zig_i128 lhs, uint8_t rhs) {
diff --git a/stage1/zig.h b/stage1/zig.h
index 188800cde497..8abc14d9d365 100644
--- a/stage1/zig.h
+++ b/stage1/zig.h
@@ -1504,8 +1504,16 @@ static inline zig_u128 zig_shl_u128(zig_u128 lhs, uint8_t rhs) {
 }
 
 static inline zig_i128 zig_shr_i128(zig_i128 lhs, uint8_t rhs) {
+    // This works around a GCC miscompilation, but it has the side benefit of
+    // emitting better code. It is behind the `#if` because it depends on
+    // arithmetic right shift, which is implementation-defined in C, but should
+    // be guaranteed on any GCC-compatible compiler.
+#if defined(zig_gnuc)
+    return lhs >> rhs;
+#else
     zig_i128 sign_mask = lhs < zig_make_i128(0, 0) ? -zig_make_i128(0, 1) : zig_make_i128(0, 0);
     return ((lhs ^ sign_mask) >> rhs) ^ sign_mask;
+#endif
 }
 
 static inline zig_i128 zig_shl_i128(zig_i128 lhs, uint8_t rhs) {
-- 
2.52.0

